src/Microsoft.ML.Transforms/Text/WordEmbeddingsTransform.cs

```diff
@@ -175,110 +263,145 @@ private WordEmbeddingsTransform(IHost host, ModelLoadContext ctx, IDataView inpu
             ctx.SetVersionInfo(GetVersionInfo());
 
-            SaveBase(ctx);
+            base.SaveColumns(ctx);
             ctx.Writer.WriteBoolByte(_customLookup);
             if (_customLookup)
                 ctx.SaveString(_modelFileNameWithPath);
             else
                 ctx.Writer.Write((uint)_modelKind);
         }
 
-        protected override ColumnType GetColumnTypeCore(int iinfo)
+        protected override IRowMapper MakeRowMapper(ISchema schema)
+           => new Mapper(this, schema);
+
+        protected override void CheckInputColumn(ISchema inputSchema, int col, int srcCol)
         {
-            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
-            return _outputType;
+            var colType = inputSchema.GetColumnType(srcCol);
+            if (!(colType is VectorType && colType.ItemType.IsText))
+                throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", ColumnPairs[col].input, "Text", inputSchema.GetColumnType(srcCol).ToString());
         }
 
-        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
+        private sealed class Mapper : MapperBase
         {
-            Host.AssertValue(ch);
-            ch.AssertValue(input);
-            ch.Assert(0 <= iinfo && iinfo < Infos.Length);
-            disposer = null;
+            private readonly WordEmbeddingsTransform _parent;
+            private readonly Model _currentVocab;
+            private readonly VectorType _outputType;
+            private readonly ColumnType[] _inputTypes;
 
-            var info = Infos[iinfo];
-            if (!info.TypeSrc.IsVector)
+            public Mapper(WordEmbeddingsTransform parent, ISchema inputSchema)
+                : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
             {
-                throw Host.ExceptParam(nameof(input),
-                    "Text input given, expects a text vector");
-            }
-            return GetGetterVec(ch, input, iinfo);
-        }
+                Contracts.CheckValue(inputSchema, nameof(inputSchema));
+                Contracts.CheckValue(parent, nameof(parent));
 
-        private ValueGetter<VBuffer<float>> GetGetterVec(IChannel ch, IRow input, int iinfo)
-        {
-            Host.AssertValue(ch);
-            ch.AssertValue(input);
-            ch.Assert(0 <= iinfo && iinfo < Infos.Length);
+                _parent = parent;
+                _inputTypes = new ColumnType[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    _inputTypes[i] = inputSchema.GetColumnType(ColMapNewToOld[i]);
+                    if (!_inputTypes[i].IsVector || !_inputTypes[i].ItemType.IsText)
+                        throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", _parent.ColumnPairs[i].input, new VectorType(TextType.Instance).ToString(), _inputTypes[i].ToString());
```

Comment:
this  is basically _parent.CheckInputColumn

Reply:
at least it should be, and you have slightly different implementations, which I would suggest to move into one.

---
In reply to: [218255515](https://github.com/dotnet/machinelearning/pull/928#discussion_r218255515) [](ancestors = 218255515)

Reply:
Actually this is the same check. I am wondering if we need to do here as well because  _parent.CheckInputColumn does it already?

---
In reply to: [218255515](https://github.com/dotnet/machinelearning/pull/928#discussion_r218255515) [](ancestors = 218255515)